packages/pwa-kit-react-sdk/src/ssr/universal/hooks/storefront-preview/index.js

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2023, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import {useEffect} from 'react'
+
+/** Origins that are allowed to run Storefront Preview. */
+const TRUSTED_ORIGINS = [
+    'https://runtime.commercecloud.com',
+    'https://runtime-admin-staging.mobify-storefront.com',
+    'https://runtime-admin-preview.mobify-storefront.com'
+]
+
+/** Origin used for local Runtime Admin. */
+const DEVELOPMENT_ORIGIN = 'http://localhost:4000'
+
+/** Detects whether the storefront is running in an iframe. */
+const detectInIframe = () => typeof window !== 'undefined' && window.parent !== window.self
+
+/** Gets the parent origin when running in an iframe. */
+const getParentOrigin = () => {
+    if (detectInIframe()) {
+        if (window.location.ancestorOrigins) return window.location.ancestorOrigins[0]
+        // ancestorOrigins does not exist in Firefox, so we use referrer as a fallback
+        if (document.referrer) return new URL(document.referrer).origin
+    }
+}
+
+const isParentOriginTrusted = (parentOrigin) => {
+    return window.location.hostname === 'localhost'
+        ? parentOrigin === DEVELOPMENT_ORIGIN // Development
+        : TRUSTED_ORIGINS.includes(parentOrigin) // Production
+}
+
+/** Detects whether the storefront is running in an iframe as part of Storefront Preview. */
+const detectStorefrontPreview = () => {
+    const parentOrigin = getParentOrigin()
+    return Boolean(parentOrigin) && isParentOriginTrusted(parentOrigin)
+}
+
+/** Returns the URL to load the Storefront Preview client script from the parent origin. */
+const getClientScript = () => {
+    const parentOrigin = getParentOrigin() ?? 'https://runtime.commercecloud.com'
+    return parentOrigin === DEVELOPMENT_ORIGIN
+        ? `${parentOrigin}/mobify/bundle/development/static/storefront-preview.js`
+        : `${parentOrigin}/cc/b2c/preview/preview.client.js`
+}
+
+/**
+ * Initializes the `window.STOREFRONT_PREVIEW` object and returns a component that will create the
+ * connection to Runtime Admin.
+ * @param {object} customizations - Customizations to the Storefront Preview behavior
+ * @param {boolean} enabled - Whether Storefront Preview is enabled. Defaults to detecting whether
+ * the storefront is framed by Runtime Admin.
+ */
+export const useStorefrontPreview = ({enabled = detectStorefrontPreview()} = {}) => {
+    // Inject the client script. We are not using a React component because do not want to make
+    // storefronts responsible for including the script themselves.
```

Comment:
I agree that we don't want the storefronts to be responsible for including the scripts.

But the implementation is anti-react, we should try to avoid manipulating the DOM directly (bypass react).

We can make a react component and keep the responsibility of injecting the script inside the SDK.

Reply:
Today, we have many SDK components.

<img width="1351" alt="Screenshot 2023-09-05 at 10 46 16 AM" src="https://github.com/SalesforceCommerceCloud/pwa-kit/assets/10948652/64fc0710-003e-4186-afff-bfad10330e1a">

Imagine adding a component `withStorefrontPreview` and wrap the app with it in the rendering pipeline.